.NET/Microsoft.Recognizers.Text/Recognizer.cs

```diff
@@ -1,47 +1,47 @@
-        {
-            modelContainer.RegisterModel(culture, type, model, options);
-        }
-
-        protected void RegisterModel(string culture, string options, Dictionary<Type, IModel> models)
-        {
-            modelContainer.RegisterModel(culture, models, options);
-        }
-
-        public IModel GetModel<TModel>(string culture, bool fallbackToDefaultCulture = true, string options = NoneOptions)
-        {
-            return modelContainer.GetModel<TModel>(culture, fallbackToDefaultCulture, options);
-        }
-
-        public bool TryGetModel<TModel>(string culture, out IModel model, bool fallbackToDefaultCulture = true, string options = NoneOptions)
-        {
-            return modelContainer.TryGetModel<TModel>(culture, out model, fallbackToDefaultCulture, options);
-        }
-
-        protected bool ContainsModels()
-        {
-            return modelContainer.ContainsModels();
-        }
-
-        public bool ContainsModel<TModel>(string culture, bool fallbackToDefaultCulture = true, string options = NoneOptions)
-        {
-            return modelContainer.ContainsModel<TModel>(culture, fallbackToDefaultCulture, options);
-        }
-
-        protected IModel GetSingleModel<TModel>()
-        {
-            return modelContainer.GetSingleModel<TModel>();
-        }
-    }
-}
+﻿using Microsoft.Recognizers.Text.Utilities;
+using System;
+
+namespace Microsoft.Recognizers.Text
+{
+    public abstract class Recognizer<TModelOptions> where TModelOptions : struct
+    {
+        private readonly ModelFactory<TModelOptions> factory;
+
+        public string TargetCulture { get; private set; }
+
+        public TModelOptions Options { get; private set; }
+
+        protected Recognizer(string targetCulture, TModelOptions options, bool lazyInitialization)
+        {
+            this.Options = options;
+            this.TargetCulture = targetCulture;
+
+            this.factory = new ModelFactory<TModelOptions>();
+            InitializeConfiguration();
+
+            if (!lazyInitialization)
+            {
+                this.InitializeModels(targetCulture, options);
+            }
+        }
+        
+        protected T GetModel<T>(string culture, bool fallbackToDefaultCulture) where T : IModel
+        {
+            return this.factory.GetModel<T>(culture ?? TargetCulture, fallbackToDefaultCulture, Options);
+        }
+
+        protected void RegisterModel<T>(string culture, Func<TModelOptions, IModel> modelCreator)
+        {
+            this.factory.Add((culture, typeof(T)), modelCreator);
+        }
+
+        protected abstract void InitializeConfiguration();
+
+        private void InitializeModels(string targetCulture, TModelOptions options)
+        {
+            this.factory.InitializeModels(targetCulture, options);
+        }
+
+        public static TModelOptions GetOption(int value) => EnumUtils.Convert<TModelOptions>(value);
```

Comment:
Shouldn't it be GetOptions?

Reply:
Since ` TModelOptions`  can be decorated with `[Flags]` attribute. Yes, the method should be named ` GetOptions`.